### PR TITLE
Add origin country rule

### PR DESCRIPTION
**Pull request by tm-kn**
This adds detection of user's country. It uses Cloudflare, CloudFront and GeoIP2 to detect user's IP address.

This PR adds `pycountry` dependency, I did not pin a version since its API is fairly stable (by the looks of it) and a lot of project use that as a dependency, so pinning a single version may cause incompatibilities of other libraries/applications.

### CODE REVIEW

File: src/wagtail_personalisation/utils.py [hunk 1]
```diff
@@ -116,3 +116,11 @@ def can_delete_pages(pages, user):
         if not variant.permissions_for_user(user).can_delete():
             return False
     return True
+
+
+def get_client_ip(request):
+    try:
+        x_forwarded_for = request.META['HTTP_X_FORWARDED_FOR']
+        return x_forwarded_for.split(',')[-1].strip()
```
Comment (m1kola) [inline src/wagtail_personalisation/utils.py]: I wonder if we should make it configurable? In some cases user's IP will be second to the last. For example, google cloud load balancer puts proxies at the end.

Comment (tm-kn) [reply src/wagtail_personalisation/utils.py]: We probably should, yes.

File: tests/unit/test_rules_country_origin.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+def test_get_cloudflare_country_with_no_header(rf):
+    segment = SegmentFactory(name='Test segment')
+    rule = OriginCountryRuleFactory(segment=segment, country='GB')
+    request = rf.get('/')
+    assert 'HTTP_CF_IPCOUNTRY' not in request.META
+    assert rule.get_cloudflare_country(request) is None
+
+
+@pytest.mark.django_db
+def test_get_cloudfront_countryo_with_header(rf):
```
Comment (jberghoef) [inline tests/unit/test_rules_country_origin.py]: Typo in the word `country`.

File: tests/unit/test_rules_country_origin.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+from wagtail_personalisation.rules import get_geoip_module
+
+
+@pytest.mark.django_db
+def test_get_cloudflare_country_with_header(rf):
+    segment = SegmentFactory(name='Test segment')
+    rule = OriginCountryRuleFactory(segment=segment, country='GB')
+    request = rf.get('/')
+    request.META['HTTP_CF_IPCOUNTRY'] = 'PL'
+    assert rule.get_cloudflare_country(request) == 'pl'
```
Comment (jberghoef) [inline tests/unit/test_rules_country_origin.py]: Is the `test_user` function evaluated in these tests?

Comment (tm-kn) [reply tests/unit/test_rules_country_origin.py]: https://github.com/wagtail/wagtail-personalisation/blob/6b3f90587577d6809723a6c9d33ace83bda978e9/tests/unit/test_rules_country_origin.py#L100-L107

That's true, I don't evaluate what value it returns. Will fix.

File: docs/default_rules.rst [hunk 4]
```diff
@@ -131,3 +131,34 @@ Is logged in        Whether the user is logged in or logged out.
...
+==================  ==========================================================
+
+You must have one of the following configurations set up in order to
+make it work.
+
+- Cloudflare IP Geolocation - ``cf-ipcountry`` HTTP header set with a value of
+  the alpha-2 country format.
+- CloudFront Geo-Targeting - ``cloudfront-viewer-country`` header set with a
+  value of the alpha-2 country format.
+- The last fallback is to use GeoIP2 module that is included with Django. It
```
Comment (tomdyson) [inline docs/default_rules.rst]: It -> This

File: src/wagtail_personalisation/utils.py (hunk 1 above)
Comment (tm-kn) [reply src/wagtail_personalisation/utils.py]: I've changed this. Can you re-review please?

File: tests/unit/test_rules_country_origin.py (hunk 3 above)
Comment (tm-kn) [reply tests/unit/test_rules_country_origin.py]: Hi @jberghoef, I added more test cases with the `test_user` method. Can you look and review again please? Thanks!

File: src/wagtail_personalisation/rules.py [hunk 5]
```diff
@@ -408,3 +430,65 @@ def description(self):
...
+
+
+class OriginCountryRule(AbstractBaseRule):
+    """
+    Test user against the country or origin of their request.
+
+    Using this rule requires setting up GeoIP2 on Django or using
+    CloudFlare or CloudFront geolocation detection.
+    """
+    country = models.CharField(max_length=2, choices=COUNTRY_CHOICES,
```
[the remark this answers is not in the document]
Comment (tm-kn) [reply src/wagtail_personalisation/rules.py]: Done.

**Pull request merged**